{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phyloseq and Microbiome analysis in R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are a number of packages developed in R that make microbiome analysis easy and produce great figures.\n",
    "There is at times a lot of overlap between this analysis and QIIME - and the choice is yours.\n",
    "\n",
    "Personally I think that R allows alot more freedom and is more aesthetically pleasing, plus there are lots of things that cannot be done in QIIME that can be done in R.\n",
    "Having said that it does require some knowledge of R, and however this notebook tries to step you through what you need to know.\n",
    "\n",
    "Two packages in particular are useful for microbiome analysis - the microbiome package builds on phyloseq, and you may find you don't need the microbiome package.\n",
    "\n",
    "[Phyloseq](http://joey711.github.io/phyloseq/)\n",
    "\n",
    "[Microbiome](http://microbiome.github.io/microbiome/)\n",
    "\n",
    "As always there is more than one way to do things, this phyloseq import will follow the basic import tutorial, in other words importing files individually to create a phyloseq object. However, you may like to follow their tutorials on the QIIME, or mothur input. I do find that as these programs change over time there can be a delay in updating the latest data import workflow, hence why I have gone with the metheod of phyloseq-ize Data already in R that will work no matter what updates are made to QIIME, usearch or other pipelines you may use."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "** Libraries**\n",
    "\n",
    "*Note: You will need to install the library first if you haven't already using* `install.packages`\n",
    "\n",
    "Load the libraries\n",
    "\n",
    "    library(\"phyloseq\")\n",
    "    library(\"tidyverse\")\n",
    "    library(\"ape\")\n",
    "    library(\"plyr\")\n",
    "    library(\"readr\")\n",
    "    library(\"microbiome\")\n",
    "    library(\"lattice\")\n",
    "    library(\"colorspace\")\n",
    "    library(\"RColorBrewer\")\n",
    "    library(\"vegan\")\n",
    "    library(\"microbiome\")\n",
    "    library(\"jsonlite\")\n",
    "    library(\"xlsx\")\n",
    "    \n",
    "Remove variables in current environment\n",
    "\n",
    "    rm(list=ls())\n",
    "\n",
    "Set working directory\n",
    "\n",
    "    setwd(\"~/tick_meta_analysis/data\")\n",
    "    \n",
    "*_Remember the useful package for data management in R_ [ProjectTemplate](http://projecttemplate.net/getting_started.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Importing data and creating a physeq object\n",
    "\n",
    "Next you can set your working directory and import your OTU table (from the usearch pipeline) and your taxonomy table (from QIIME). I recommend you copy and data the files from your Usearch/QIIME directory output to the data file in the R management directory\n",
    "    \n",
    "    setwd(\"~/tick_meta_analysis/data\")\n",
    "    \n",
    "Import OTU table taking note of the following:\n",
    "- Rows = taxa (otus) and columns = samples. The sample names should already be matched to your metadata if you following the QIIME analysis to generate a feature table correctly.\n",
    "- Order the OTU column in ascending order _to do this you may need to delete the letters 'OTU', which can be done in excel with a `control-H` command._\n",
    "- The first column with the OTU number must then be deleted, this can get confusing so it is best to take this document as a copy so that you do not save over the table with the OTU numbers.\n",
    "- Ensure that the column headings (sample names) are in the same order as what appears in the metadata sheet. In excel you may like to copy and paste the data using the transpose option to order the sample column before reverting it back.\n",
    "- Save as `csv` file\n",
    "\n",
    "```\n",
    "uparse_otus <- read_csv(\"~/tick_meta_analysis/data/otu_table.csv\")\n",
    "otumat <- as.matrix(uparse_otus)\n",
    "rownames(otumat) <- paste0(\"OTU\", 1:nrow(otumat))\n",
    "colnames(otumat) <- paste0(\"Sample\", 1:ncol(otumat))\n",
    "otumat\n",
    "```\n",
    "\n",
    "Import the taxonomy table taking note of the following:\n",
    "- Rows = taxa and columns = taxa heirachy\n",
    "- The first column with the OTU number must then be deleted, this can get confusing so it is best to take this document as a copy so that you do not save over the table with the OTU numbers.\n",
    "- Save as `csv` file\n",
    "\n",
    "```\n",
    "colnames(taxmat) <- c(\"Kingdom\", \"Phylum\", \"Class\", \"Order\", \"Family\", \"Genus\", \"Species\")\n",
    "taxonomy <- read_csv(\"~/tick_meta_analysis/data/taxonomy.csv\")\n",
    "taxmat <- as.matrix(taxonomy)\n",
    "rownames(taxmat) <- rownames(otumat)\n",
    "colnames(taxmat) <- c(\"Kingdom\", \"Phylum\", \"Class\", \"Order\", \"Family\", \"Genus\", \"Species\")\n",
    "taxmat\n",
    "```\n",
    "\n",
    "Next wcan check the class of the `otumat` and `taxmat` objects, they **MUST** be in matrix format. Then we can great a phyloseq object called `physeq` from the otu and taxonomy tables and check the sample names.\n",
    "```\n",
    "class(otumat)\n",
    "class(taxmat)\n",
    "OTU = otu_table(otumat, taxa_are_rows = TRUE)\n",
    "TAX = tax_table(taxmat)\n",
    "OTU\n",
    "TAX\n",
    "physeq = phyloseq(OTU, TAX)\n",
    "physeq\n",
    "sample_names(physeq)\n",
    "```\n",
    "\n",
    "**There are a few options for the tree import**- you can create a random tree using the `ape` package, however as this only using taxonomy information it is not as robust as other methods that use the sequence information, such as that generated by QIIME or USEARCH.\n",
    "\n",
    "```\n",
    "random_tree = rtree(ntaxa(physeq), rooted=TRUE, tip.label=taxa_names(physeq))\n",
    "plot(random_tree)\n",
    "```\n",
    "\n",
    "To import your usearch tree\n",
    "```\n",
    "MyTree <- read.tree(\"otus.tree\")\n",
    "```\n",
    "To import your QIIME2 tree\n",
    "```\n",
    "MyTree2 <- read.tree(\"rooted-tree.nwk\")\n",
    "```\n",
    "\n",
    "Now you can merge your data to create a final phyloseq object\n",
    "```\n",
    "physeq1 = merge_phyloseq(physeq, sampledata, random_tree)\n",
    "physeq1\n",
    "```\n",
    "\n",
    "You should end up with the following output after `physeq1`\n",
    "\n",
    ">```\n",
    "phyloseq-class experiment-level object\n",
    "otu_table()   OTU Table:         [ 15985 taxa and 1390 samples ]\n",
    "sample_data() Sample Data:       [ 1390 samples by 13 sample variables ]\n",
    "tax_table()   Taxonomy Table:    [ 15985 taxa by 7 taxonomic ranks ]\n",
    "phy_tree()    Phylogenetic Tree: [ 15985 tips and 15984 internal nodes ]\n",
    ">```\n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic commands to check your physeq object\n",
    "```\n",
    "nsamples(physeq1)\n",
    "ntaxa(physeq1)\n",
    "sample_names(physeq1)[1:5]\n",
    "rank_names(physeq1)\n",
    "sample_variables(physeq1)\n",
    "otu_table(physeq1)[1:5, 1:5]\n",
    "tax_table(physeq1)[1:5, 1:4]\n",
    "myTaxa = names(sort(taxa_sums(physeq1), decreasing = TRUE)[1:10])\n",
    "```\n",
    "\n",
    "\n",
    "## Pre-processing data\n",
    "The phyloseq package also includes functions for filtering, subsetting, and merging abundance data. Filtering in phyloseq is designed in a modular fashion similar to the approach in the genefilter package. This includes the `prune_taxa` and  `prune_samples` methods for directly removing unwanted indices, as well as the `filterfun_sample` and `genefilter_sample` functions for building arbitrarily complex sample-wise filtering criteria, and the `filter_taxa` function for taxa-wise filtering. \n",
    "\n",
    "##### Transforming and filtering data\n",
    "In the following example, the Physeq1 data is first transformed to relative abundance, creating the new PHr object, which is then filtered such that only OTUs with a mean greater than 10^-5 are kept.\n",
    "```\n",
    "PHr  = transform_sample_counts(physeq1, function(x) x / sum(x) )\n",
    "PHfr = filter_taxa(PHr, function(x) mean(x) > 1e-5, TRUE)\n",
    "```\n",
    "Remeber you can check the number of taxa at any stage by\n",
    "```\n",
    "ntaxa(PHr)\n",
    "ntaxa(PHfr)\n",
    "```\n",
    "This results in a highly-subsetted object, `PHfr`, containing just 4624 of the original ~19216 OTUs. Note that in both lines we have provided a custom function for transformation and filtering, respectively.\n",
    "\n",
    "##### Subsetting data\n",
    "Subset import taxa, and set threshold limites, e.g. >20 reads.\n",
    "\n",
    "```\n",
    "c_Alphaprob = subset_taxa(physeq1, Class==\"c__Alphaproteobacteria\")\n",
    "c_Alphaprob = prune_samples(sample_sums(c__Alphaprob)>=20, c__Alphaprob)\n",
    "\n",
    "g_Borrelia = subset_taxa(physeq1, Genus==\"Borrelia\")\n",
    "g_Borrelia = prune_samples(sample_sums(g_Borrelia)>=1, g_Borrelia)\n",
    "\n",
    "f_Bartonellaceae = subset_taxa(physeq1, Family==\"Bartonellaceae\")\n",
    "f_Bartonellaceae = prune_samples(sample_sums(f_Bartonellaceae)>=1, f_Bartonellaceae)\n",
    "```\n",
    "\n",
    "#### Plotting subsets\n",
    "```\n",
    "title = \"plot_bar; Alphaproteobacteria-only\"\n",
    "plot_bar(c__Alphaprob, \"Genus_species\", title=title)\n",
    "```\n",
    "\n",
    "## Rarefying data\n",
    "\n",
    "<div class=\"alert alert-warning\"> Please note that the authors of phyloseq do not advocate using this as a normalization procedure, despite its recent popularity. See McMurdie and Holmes 2015 PLoS Comput. Biol. or \"?rarefy_even_depth\" in your R environment </div>\n",
    "\n",
    "You will need to experiment with this line of code in order for it to make biological sense to your data set. We recommend reading the reference above about Rarefying microbial datasets.\n",
    "\n",
    "```\n",
    "physeq1.rarified <- rarefy_even_depth(physeq1)\n",
    "ntaxa(physeq1.rarified)\n",
    "nsamples(physeq1.rarified)\n",
    "rm(physeq1.rarified)\n",
    "physeq1.rarified <- rarefy_even_depth(physeq1, rngseed = 1000)\n",
    "ntaxa(physeq1.rarified)\n",
    "nsamples(physeq1.rarified)\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
